### ISSUE TITLE
Use Go boring source archives, not Git branches?

### ISSUE THREAD

**Issue by dagood** [microsoft/go#434]
https://groups.google.com/g/golang-dev/c/cyN-u0b-06w/m/vhInf1u9EgAJ?utm_medium=email&utm_source=footer

> The VERSION file is removed because these are development branches. They are definitely weird development branches because releases are cut from them, but we just adopted the process we already had.
> 
> The RELEASES file is still being updated on the main dev.boringcrypto branch. https://github.com/golang/go/blob/dev.boringcrypto/misc/boring/RELEASES
> 
> You should not consider the existence of dev.boringcrypto.go1.X branches to be stable, they are just an implementation detail of how the releases are generated (which are also not officially supported, but will be more reliable than the git branches).

This suggests we should be watching that `RELEASES` file on the non-release-specific dev.boringcrypto branch and using source tarballs from there rather than using Git. The source archives in the RELEASES file actually do have VERSION files in them, unlike the branches.

On the patch/build/test side, this doesn't require much change: extract the tarball into `go/` instead of using the submodule.

Infra would need more work:
* Mirroring the source archives. (We have an internal Git repository for mirroring right now. Can we shove the source into that and keep using a submodule? What about anonymous access?)
* Auto-update needs to be highly specialized for the boring branches in particular. Completely new source retrieval process.
* When resolving patches, how do we do that reasonably without Git history of what happened in the upstream boring branch?
  * Would it be reasonable to hybridize: use the Git branch for development and patch fixups, then switch over to the source tarball once upstream releases one? Keep both approaches in our branch and use flags to switch from one source code acquisition method vs. the other?

**Comment by jaredpar** [comment]
Other than the `VERSION` file is there really any diff between the Git code and the source tarball?

**Comment by dagood** [comment]
> Other than the `VERSION` file is there really any diff between the Git code and the source tarball?

Tried out a diff on the raw source vs. the tarball: https://gist.github.com/dagood/a54cbb271871750cd117372845f3375b

Some Git/GitHub-related files are removed, and the VERSION file is added--that's it. I think this is all done by this part of the release tool:
https://github.com/golang/build/blob/2d45280fb4864d597e912aef59d65c6bc2f3156f/cmd/release/release.go#L390-L399

Reasons this situation isn't great:
* Without upstream's VERSION file, we have to maintain VERSION files ourselves. (One in each Boring branch.)
* We're being actively warned that the boring branches should not be treated as reliable. Does that mean they'll disappear, get force pushed, or maybe more diffs will appear between the branch content and the released tarballs? None of this seems *likely* to me, but possible.

We did get another bit of detail on that thread today:

> With our current process, the commits will always be in the repository by the time they are checked into the RELEASES file.

Timing of releases was a pretty big concern I had, so it's good to read that upstream isn't *planning* to withhold Git pushes until sometime after the release. The scenario I was imagining: a CVE fix in 1.17 that we need to update 1.17-fips to release, so we need upstream to merge the underlying boring branch. But upstream only publishes a boring source tarball, no Git commit! Our submodule wouldn't be able to get the source. We would need to do some quick changes to get it out the door.

---

A potential solution for all these scenarios that I somehow didn't think of earlier is to generate a patch file that changes the last-released Git commit into the source code in the released tarball. I think that's a fine stopgap.

I think if something like this ever actually happens, we can evaluate at that point whether we think it'll happen again and decide whether to switch over to tarballs then. (Closing.)
